File: python_plots_generator/app/Server/DTOS/DoubleLinearDTO.py
from flask_restx import fields, Namespace

from app.Config.Constants import FORMATTERS_LIST

"""

    {
        "src": {
            "path": "spark/temp/evol/cadiz/evol",
            "axis": {
                "x": {
                    "name": "date",
                    "format": "timestamp"
                },
                "y_1": {
                    "name": "temp_daily_avg"
                },
                "y_2": {
                    "name": "temp_daily_avg"
                },
            }
        },
        "dest" : {
            "path": "temp/evol/cadiz/evol",
            "filename": "plot_test",
            "export_png": true
        },
        "style": {
            "lettering": {
                "title": "TITLE TEST",
                "subtitle": "SUBTITLE TEST",
                "x_label": "Date",
                "y_1_label": "Var (u)",
                "y_2_label": "Var (u)"
            },
            "figure_1": {
                "name": "var",
                "color": "#4169e1"
            },
            "figure_2": {
                "name": "var",
                "color": "#4169e1"
            },
            "margin": {
                "left": 120,
                "right": 120,
                "top": 100,
                "bottom": 100
            },
            "legend": {
                "y_offset": -0.05
            }
        }
    }


"""

class DoubleLinearDTO:
    def __init__(self, ns: Namespace):
        self.post_input = _create_input_post_dto(ns)


def _create_input_post_dto(ns: Namespace):
    return ns.model('DoubleLinearInput', {
        'src': fields.Nested(ns.model('DoubleLinearSrc', {
            'path': fields.String(required=True, description="Relative route to data"),
            'axis': fields.Nested(ns.model('DoubleLinearSrcAxis', {
                'x': fields.Nested(ns.model('DoubleLinearSrcAxisX', {
                    'name': fields.String(reqired=True, description="X Column name"),
                    'format': fields.String(reqired=False, description="Column format", enum=FORMATTERS_LIST),
                }), required=True),
                'y_1': fields.Nested(ns.model('DoubleLinearSrcAxisY1', {
                    'name': fields.String(reqired=True, description="Y 1 Column name"),
                }), required=True),
                'y_2': fields.Nested(ns.model('DoubleLinearSrcAxisY2', {
                    'name': fields.String(reqired=True, description="Y 2 Column name"),
                }), required=True)
            }), required=True)
        })),
        'dest': fields.Nested(ns.model('DoubleLinearDest', {
            'path': fields.String(required=True, description="Relative output route"),
            'filename': fields.String(required=True, description="Filename of the chart"),
            'export_png': fields.Boolean(required=True, description="Export png"),
        }), required=True),
        'style': fields.Nested(ns.model('DoubleLinearStyle', {
            'lettering': fields.Nested(ns.model('DoubleLinearStyleLettering', {
                'title': fields.String(required=True, description="Title of the chart"),
                'subtitle': fields.String(required=False, description="Subtitle of the chart"),
                'x_label': fields.String(required=True, description="X axis label"),
                'y_1_label': fields.String(required=True, description="Y 1 axis label"),
                'y_2_label': fields.String(required=True, description="Y 2 axis label"),
            }), required=True),
            'figure_1': fields.Nested(ns.model('DoubleLinearStyleFigure1', {
                'name': fields.String(required=True, description="Figure 1 name"),
                'color': fields.String(required=True, description="Figure 1 color"),
            }), required=True),
            'figure_2': fields.Nested(ns.model('DoubleLinearStyleFigure2', {
                'name': fields.String(required=True, description="Figure 2 name"),
                'color': fields.String(required=True, description="Figure 2 color"),
            }), required=True),
            'margin': fields.Nested(ns.model('DoubleLinearStyleMargin', {
                'left': fields.Float(required=True, description="Margin left"),
                'right': fields.Float(required=True, description="Margin right"),
                'top': fields.Float(required=True, description="Margin top"),
                'bottom': fields.Float(required=True, description="Margin bottom")
            }), required=False),
            'legend': fields.Nested(ns.model('DoubleLinearStyleLegend', {
                'y_offset': fields.Float(required=True, description="Legend Y axis offset"),
            }), required=False),
        }), required=True)
    })

File: scala_big_data_core/src/main/scala/Config/PlotGenerationConf/Execution/DTO/DoubleLinearDTO.scala
package Config.PlotGenerationConf.Execution.DTO

import upickle.default.{ReadWriter, macroRW}
import upickle.implicits.key

case class DoubleLinearDTOAxisComponent(
  @key("name") name: String,
  @key("format") format: Option[String]
)
object DoubleLinearDTOAxisComponent { implicit val rw: ReadWriter[DoubleLinearDTOAxisComponent] = macroRW }

case class DoubleLinearDTOAxis(
  @key("x") x: DoubleLinearDTOAxisComponent,
  @key("y1") y1: DoubleLinearDTOAxisComponent,
  @key("y2") y2: DoubleLinearDTOAxisComponent
)
object DoubleLinearDTOAxis { implicit val rw: ReadWriter[DoubleLinearDTOAxis] = macroRW }

case class DoubleLinearDTOSrc(
  @key("path") path: String,
  @key("axis") axis: DoubleLinearDTOAxis
)
object DoubleLinearDTOSrc { implicit val rw: ReadWriter[DoubleLinearDTOSrc] = macroRW }

case class DoubleLinearDTODest(
  @key("path") path: String,
  @key("filename") filename: String,
  @key("export_png") exportPng: Boolean
)
object DoubleLinearDTODest { implicit val rw: ReadWriter[DoubleLinearDTODest] = macroRW }

case class DoubleLinearDTOLettering(
  @key("title") title: String,
  @key("subtitle") subtitle: String,
  @key("x_label") xLabel: String,
  @key("y1_label") y1Label: String,
  @key("y2_label") y2Label: String
)
object DoubleLinearDTOLettering { implicit val rw: ReadWriter[DoubleLinearDTOLettering] = macroRW }

case class DoubleLinearDTOFigure(
  @key("name") name: String,
  @key("color") color: String
)
object DoubleLinearDTOFigure { implicit val rw: ReadWriter[DoubleLinearDTOFigure] = macroRW }

case class DoubleLinearDTOMargin(
  @key("left") left: Float,
  @key("right") right: Float,
  @key("top") top: Float,
  @key("bottom") bottom: Float
)
object DoubleLinearDTOMargin { implicit val rw: ReadWriter[DoubleLinearDTOMargin] = macroRW }

case class DoubleLinearDTOLegend(
  @key("y_offset") yOffset: Float
)
object DoubleLinearDTOLegend { implicit val rw: ReadWriter[DoubleLinearDTOLegend] = macroRW }

case class DoubleLinearDTOStyle(
  @key("lettering") lettering: DoubleLinearDTOLettering,
  @key("figure_1") figure1: DoubleLinearDTOFigure,
  @key("figure_2") figure2: DoubleLinearDTOFigure,
  @key("margin") margin: Option[DoubleLinearDTOMargin],
  @key("legend") legend: Option[DoubleLinearDTOLegend]
)
object DoubleLinearDTOStyle { implicit val rw: ReadWriter[DoubleLinearDTOStyle] = macroRW }

case class DoubleLinearDTO (
  @key("src") src: DoubleLinearDTOSrc, 
  @key("dest") dest: DoubleLinearDTODest, 
  @key("style") style: DoubleLinearDTOStyle
)
object DoubleLinearDTO { implicit val rw: ReadWriter[DoubleLinearDTO] = macroRW }
